src/v/storage/segment_set.cc

```diff
@@ -159,27 +160,55 @@ unsafe_do_recover(segment_set&& segments, ss::abort_source& as) {
             return std::move(segments);
         }
         segment_set::underlying_t good = std::move(segments).release();
+        absl::btree_set<segment*> to_recover_set;
+        to_recover_set.insert(
+          good.back().get()); // always recover the last segment
+
+        for (size_t i = 0; i < good.size(); ++i) {
+            auto& s = *good[i];
+            if (i > 0) {
+                auto& prev = *good[i - 1];
+
+                if (prev.offsets().dirty_offset >= s.offsets().base_offset) {
```

Comment:
I think we can make this a more comprehensive check by also checking that dirty_offset >= max_offset here, in case of other offset bug -- it's a great position to check all the things we possibly can, while are still early enough in startup to rewrite indices.

Reply:
> also checking that dirty_offset >= max_offset here

Can you elaborate? I thought they are essentially the same, i.e. dirty_offset gets assigned index.max_offset after the index is hydrated.

Reply:
Right, they should be the same - I'm basically suggesting some paranoia: this is a convenient place to check all the invariants we assume about the state of the index/segment variables.